R/slide.R

```diff
@@ -8,27 +8,37 @@
 #' @param .f Function, formula, or missing; together with `...` specifies the
 #'   computation to slide. To "slide" means to apply a computation within a
 #'   sliding (a.k.a. "rolling") time window for each data group. The window is
-#'   determined by the `before` and `after` parameters described below. One time
-#'   step is typically one day or one week; see details for more explanation. If
-#'   a function, `.f` must take a data frame with the same column names as the
-#'   original object, minus any grouping variables, containing the time window
-#'   data for one group-`.ref_time_value` combination; followed by a one-row
-#'   tibble containing the values of the grouping variables for the associated
-#'   group; followed by any number of named arguments. If a formula, `.f` can
-#'   operate directly on columns accessed via `.x$var` or `.$var`, as in
-#'   `~mean(.x$var)` to compute a mean of a column `var` for each
+#'   determined by the `.window_size` and `.align` parameters, see the details
+#'   section for more. If a function, `.f` must have the form `function(x, g, t,
+#'   ...)`, where
+#'
+#'   - "x" is a data frame with the same column names as the original object,
+#'   minus any grouping variables, with only the windowed data for one
+#'   group-`.ref_time_value` combination
+#'   - "g" is a one-row tibble containing the values of the grouping variables
+#'   for the associated group
+#'   - "t" is the ref_time_value for the current window
+#'   - "..." are additional arguments
+#'
+#'   If a formula, `.f` can operate directly on columns accessed via `.x$var` or
+#'   `.$var`, as in `~mean(.x$var)` to compute a mean of a column `var` for each
 #'   `ref_time_value`-group combination. The group key can be accessed via `.y`.
 #'   If `.f` is missing, then `...` will specify the computation.
 #' @param ... Additional arguments to pass to the function or formula specified
 #'   via `.f`. Alternatively, if `.f` is missing, then the `...` is interpreted
 #'   as a ["data-masking"][rlang::args_data_masking] expression or expressions
 #'   for tidy evaluation; in addition to referring columns directly by name, the
 #'   expressions have access to `.data` and `.env` pronouns as in `dplyr` verbs,
-#'   and can also refer to `.x`, `.group_key`, and `.ref_time_value`. See
-#'   details.
+#'   and can also refer to `.x` (not the same as the input epi_df),
+#'   `.group_key`, and `.ref_time_value`. See details.
 #' @param .new_col_name String indicating the name of the new column that will
 #'   contain the derivative values. Default is "slide_value"; note that setting
 #'   `new_col_name` equal to an existing column name will overwrite this column.
+#' @param .complete_only Logical; if `TRUE`, only slide values that have a
```

Comment:
> different from what we get out of epi_slide_sum, which outputs rows with explicit NAs by default

have some TODOs above the consistency tests for `epi_slide/epi_slide_opt` (line ~300 in test-epi_slide.R):

```
  # These are the consistency tests between epi_slide and epi_slide_opt
  # functions. Only the specific case of .complete_only = FALSE and the opt
  # functions using na.rm = TRUE is testsed (the two options are equivalent for
  # our purposes here).
  # TODO: See if we can include the .complete_only = TRUE case in the future.
  # TODO: Add a case where the data contains NA values (not just gaps in time_value).
```

agree that it would be nice to get something consistent between the two.

> todo: note the default value of .complete_only in docs here.

done

> todo: also make the grouping demands consistent between epi_slide and epi_slide_opt; e.g., the former currently will auto-group an ungrouped epi_df, while the latter refuses to continue

will do. haven't propagate changes from epi_slide to opt yet, but definitely intended.

> note: .complete_only = TRUE seems significantly slower than FALSE

hm, we should definitely spend some time to address this (though let's do it after we settle the questions about what behaviors from epi_slide_opt we want to imitate; our decisions there might make things easier). `slider::slider_hop_index` doesn't help us unfortunately, because its `complete = TRUE` argument only applies to windows on the boundary (iirc), so I had to implement my own (and unfortunately slow) check for window completeness. one option is just to imitate epi_slide_opt and create a contiguous index by default and then we don't need to check for completeness (users can forward na.rm to their `.f` if they desire).

Reply:
update: given recent conversations, going to remove .complete_only and go with the work Nat has done in #415 that autocompletes the epi_df.